src/Microsoft.OData.Core/SelectedPropertiesNode.cs

```diff
@@ -336,11 +339,62 @@ internal IEnumerable<IEdmNavigationProperty> GetSelectedNavigationProperties(IEd
             return selectedNavigationProperties.Distinct();
         }
 
+        /// <summary>
+        /// Gets all selected structured properties at the current node.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The selected structured properties at this node level.</returns>
+        /// <remarks>Result does not include the dynamic properties, which can be retrieved by <code>GetSelectedDynamicProperties</code></remarks>
+        internal IEnumerable<IEdmProperty> GetSelectedProperties(IEdmStructuredType structuredType)
+        {
+            if (this.selectionType == SelectionType.Empty)
+            {
+                return EmptyEdmProperties;
+            }
+
+            // We cannot determine the selected properties without the user model. This means we won't be computing the missing properties.
+            if (structuredType == null)
+            {
+                return EmptyEdmProperties;
+            }
+
+            if (this.selectionType == SelectionType.EntireSubtree || this.hasWildcard)
+            {
+                return structuredType.DeclaredProperties;
+            }
+
+            Debug.Assert(this.selectedProperties != null, "selectedProperties != null");
+
+            // Get declared properties selected, and filter out unrecognized properties.
+            return this.selectedProperties.Select(structuredType.FindProperty).OfType<IEdmProperty>();
+        }
+
+        /// <summary>
+        /// Return names of dynamic properties if current structured node is open type.
+        /// </summary>
+        /// <param name="structuredType">The current structured type.</param>
+        /// <returns>The names of dynamic properties, empty if the current structured type is not open type.</returns>
+        internal IEnumerable<string> GetSelectedDynamicProperties(IEdmStructuredType structuredType)

```

Comment:
>GetSelectedDynamicProperties [](start = 37, length = 28)

Why are we breaking dynamic properties into a separate method from declared properties?  Doesn't that require us to do multiple loops through selected properties (and then, for each selected property, find out if it is declared)?

Why don't we just do a single loop through selectedProperties and process each as appropriate?

Reply:
These two methods are invoked from code segment around L2287. Regular property restoration needs to check for nullable type (otherwise no restoration); while dynamic property doesn't require such check. Therefore two passes are used here (same linear complexity of O(N)).

---
In reply to: [213517170](https://github.com/OData/odata.net/pull/1106#discussion_r213517170) [](ancestors = 213517170)